grafana/opensearch-datasource
# Service Map: Implement service map index prefetch

Split the feature into multiple PRs for ease of reviewing. The feature is not enabled by deafult but 'hidden' behind a provisional feature flag. 

In order to get stats for the service map (service info that's displayed inside a node: `avg latency`, `throughput`, `error rate`), we have to build an aggregation query on the `span` indices. The documents (spans) need to be filtered by operationNames and serviceNames, in order to get and aggregate *only* on spans that represent requests from one service to another (not internal processes). 
This reflects how the aggregations are build for the Service Map panel in Open Search. 

To get the operationNames and serviceNames, we have to first query the `service-map` indices and get operations and services from the `serviceName` and `destination.resource` fields of the index. 

**This PR implements:** 

1. detecting if the query requested from GF Open Search plugin has the serviceMap flag enabled
2. creating a separate query just for the `service-map` indices with the flag `serviceMapPrefetch`
3. adding flows that detect the `serviceMapPrefetch` flag and build the aggregations and the query
4. process the response from OpenSearch and store operation names and service names into Data Frames
5. putting those operations and services into the model of a service map Stats query and forwarding it along with the original query array (can contain other types of queries as well)
6. subsequent PRs will deal with building the stats query and processing it into final data frames that will be displayed in Grafana Node Graph visualization ([draft PR](https://github.com/grafana/opensearch-datasource/pull/363))

**Note:** This builds queries for otel + Open Search data prepper. Open Search can [store traces in raw Jaeger format](https://opensearch.org/docs/latest/observing-your-data/trace/index/#trace-analytics-with-jaeger-data), in which case the aggregations and responses should look slightly different. This is something we intend to implement in the future. 

**To test:**
1. enable `openSearchNodeGraph` feature toggle in Grafana config
2. run the Data Prepper/examples/jaeger-hotrod
3. generate some traces in the browser app
4. run the plugin in debugging mode
5. select query type: Traces and switch on the Service Map button in the query editor
6. you should be getting services and operations when you break on [this line](https://github.com/grafana/opensearch-datasource/blob/f6e1db87d81b77f8173703fa42ed8047c0c27c2e/pkg/opensearch/opensearch.go#L98)
7. nothing will show in the visualization yet, since we don't have stats processing yet
